[c10d] Make allreduce as a custom op

alanwaketan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79779
* #79777
* #79691
* #79688
* #79687
* #79686
* #79683
* #79669
* __->__ #79582

Summary:
This patch makes allreduce as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_allreduce_ops
python test/distributed/test_c10d_gloo.py -k test_allreduce_basics
...and other existing distributed tests.

Commit:
[c10d] Make allreduce as a custom op

Summary:
This patch makes allreduce as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_allreduce_ops
python test/distributed/test_c10d_gloo.py -k test_allreduce_basics
...and other existing distributed tests.

[ghstack-poisoned]

Commit:
Update on "[c10d] Make allreduce as a custom op"


Summary:
This patch makes allreduce as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_allreduce_ops
python test/distributed/test_c10d_gloo.py -k test_allreduce_basics
...and other existing distributed tests.

[ghstack-poisoned]

ezyang commented on torch/csrc/distributed/c10d/default_comm_hooks.cpp:
```diff
@@ -23,7 +28,12 @@ c10::intrusive_ptr<c10::ivalue::Future> FP16CompressCommHook::runHook(
+  auto allreduce_fut =
+      ops::allreduce(
+          c10::intrusive_ptr<ProcessGroup>::unsafe_reclaim_from_nonowning(
+              state_),
```
What's going on here?

alanwaketan replied:
I'm trying to convert a raw pointer to an intrusive_ptr. Is this the way to do so?

mrshenli replied:
curious, who owns this PG instance? I assume it is owned by Python PG object? If that's the case, will this mess up the refcnt. What happens when this tmp intrusive ptr exits scope?

alanwaketan replied:
I think PG are normally owned by a Python object. AllReduceCommHook somehow holds a ProcessGroup* instead of intrusive_ptr<ProcessGroup>. Therefore, I need to convert the raw pointer to a intrusive_ptr.

I don't believe this will mess up the refcnt. However, I actually think it's better to replace `class AllReduceCommHook : public CppCommHookInterface<ProcessGroup*>` with `class AllReduceCommHook : public CppCommHookInterface<intrusive_ptr<ProcessGroup>>`. What do you think?

mrshenli replied:
> I don't believe this will mess up the refcnt. 

If we create an intrusive ptr from the raw ptr, does this mean we have two separate entities tracking refcnt for the same raw ptr separately? One is the Python object, and another is this intrusive ptr?



> I actually think it's better to replace class AllReduceCommHook : public CppCommHookInterface<ProcessGroup*> with class AllReduceCommHook : public CppCommHookInterface<intrusive_ptr<ProcessGroup>>. What do you think?

Yep, this does sounds better to me.

alanwaketan replied:
I think for intrusive_ptr the refcnt is stored in the object (ProcessGroup) itself. Intrusive_ptr is just a way to increment/decrement the refcnt. So it shouldn't matter.

Let me make a follow up patch on changing `class AllReduceCommHook : public CppCommHookInterface<ProcessGroup*>` to `class AllReduceCommHook : public CppCommHookInterface<intrusive_ptr<ProcessGroup>>`.

mrshenli replied:
Got it. Can we also add a comment for this in the code? Thank you!

mrshenli commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -5,9 +5,9 @@
 namespace {
-c10::intrusive_ptr<ProcessGroup::Work> broadcast(
-    const c10::intrusive_ptr<ProcessGroup>& process_group,
+c10::intrusive_ptr<ProcessGroup::Work> broadcast_(
```
curious, what's the reason for adding a trailing underscore?

alanwaketan replied:
It's a way to signal this is an in-place op. I guess that's almost true for every c10d ops.

mrshenli commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -5,9 +5,9 @@
-    const c10::intrusive_ptr<ProcessGroup>& process_group,
+c10::intrusive_ptr<ProcessGroup::Work> broadcast_(
     at::TensorList tensors,
+    const c10::intrusive_ptr<ProcessGroup>& process_group,
```
Also curious about the reason for argument ordering :)

alanwaketan replied:
I think it's a convention for aten ops to have tensor-like parameters as the first argument. Especially for in-place ops, it tells that's the one get the inplace update.

mrshenli commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -18,6 +18,19 @@ c10::intrusive_ptr<ProcessGroup::Work> broadcast(
+  return process_group->allreduce(
+      tensor_vec,
+      AllreduceOptions{
+          static_cast<ReduceOp>(reduce_op),
```
any reason here for using a static cast instead of directly passing in the reduce op as an argument?

alanwaketan replied:
ReduceOp is an enum class and reduce_op is a int64_t. I think a cast is needed here.

mrshenli commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -18,6 +18,19 @@ c10::intrusive_ptr<ProcessGroup::Work> broadcast(
           root_rank, root_tensor, std::chrono::milliseconds(timeout)});
 }
 
+c10::intrusive_ptr<ProcessGroup::Work> allreduce_(
```
just a heads-up that there is an ongoing effort on consolidating `ProcessGroup::Work` with torch Future. 

cc @kumpera

alanwaketan replied:
In the last conversation, we agree that we won't have `wait()` in the graph. What we do is to fulfill the async handler immediately with the tracing proxy.

mrshenli reviewed (commented):
Thanks @alanwaketan for working on this. Is it possible to have some test coverage?

wanchaol reviewed (commented):
Overall looks great! Thanks for doing this, have a few comments and questions.

alanwaketan commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -18,6 +18,19 @@ c10::intrusive_ptr<ProcessGroup::Work> broadcast(
+c10::intrusive_ptr<ProcessGroup::Work> allreduce_(
+    at::TensorList tensors,
+    const c10::intrusive_ptr<ProcessGroup>& process_group,
+    int64_t reduce_op,
```
Yes, that's correct. I'm not sure if there is a way to encode enum for aten ops.

alanwaketan commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -18,6 +18,19 @@ c10::intrusive_ptr<ProcessGroup::Work> broadcast(
+      tensor_vec,
+      AllreduceOptions{
+          static_cast<ReduceOp>(reduce_op),
+          std::chrono::milliseconds(timeout)});
```
That's correct.

alanwaketan commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -27,8 +40,11 @@ TORCH_LIBRARY(c10d, m) {
   m.def(
-      "broadcast",
-      dispatch(c10::DispatchKey::CompositeExplicitAutograd, broadcast));
+      "broadcast_",
```
I believe this is the convention for aten ops that do inplace updates on the inputs.

alanwaketan commented on torch/csrc/distributed/c10d/Ops.cpp:
```diff
@@ -27,8 +40,11 @@ TORCH_LIBRARY(c10d, m) {
+      "broadcast_",
+      dispatch(c10::DispatchKey::CompositeExplicitAutograd, broadcast_));
+  m.def(
+      "allreduce_",
```
No, that's not reason to have _. Please refer to the other comment for the reason behind.

For the node, it should be torch.ops.c10d.allreduce_(...). I don't think there is a default parameter.

alanwaketan commented:
> Thanks @alanwaketan for working on this. Is it possible to have some test coverage?

What do you mean by test coverage? The c10d functionalities should be covered by existing tests.

Or you want me to add tests to demonstrate the ops are somehow visible to the dispatcher? I guess I can. I just don't know if that's necessary because the fact that the existing tests all pass already indirectly verifies that.

wanchaol commented:
> Or you want me to add tests to demonstrate the ops are somehow visible to the dispatcher? I guess I can. I just don't know if that's necessary because the fact that the existing tests all pass already indirectly verifies that.

One question i have is that how do we verify if we could trace the collective now? should we add some fx tracing test to ensure the collectives appear in the graph?

alanwaketan commented:
> > Or you want me to add tests to demonstrate the ops are somehow visible to the dispatcher? I guess I can. I just don't know if that's necessary because the fact that the existing tests all pass already indirectly verifies that.
> 
> One question i have is that how do we verify if we could trace the collective now? should we add some fx tracing test to ensure the collectives appear in the graph?

AOT is built on top of __torch_dispatch__. And as long as ops are made dispatcher passable, they should work with __torch_dispatch__. I guess I can add tests to have a python tensor that override __torch_dispatch__ to directly verify that. I just don't feel it's necessary. Let me know if you demand that.

wanchaol reviewed (commented):
> I guess I can add tests to have a python tensor that override torch_dispatch to directly verify that.

I see, I guess we can add those tests later in a separate PR when we necessarily need it. There's two things on top of my head and need some inputs from @mrshenli, as these might be related to the actual node appears in the IR, we should get some clarify and make them consistent:
1. about operator suffix and argument ordering: should we make the aten operator follow our [python level API](https://github.com/pytorch/pytorch/blob/09df27fe452dbc91f8a0eb069cd714deeaf12443/torch/distributed/distributed_c10d.py#L1172), or should we follow the ATen operator naming convention?
2. should we let `wait()` appear in the IR? This might be related to how the cuda stream sync works in our current tracer.

alanwaketan commented:
> 1. about operator suffix and argument ordering: should we make the aten operator follow our [python level API](https://github.com/pytorch/pytorch/blob/09df27fe452dbc91f8a0eb069cd714deeaf12443/torch/distributed/distributed_c10d.py#L1172), or should we follow the ATen operator naming convention?

I think we should follow the aten convention for the funtion schema as that will be easier for any tracer to interpret the ops. At least AOT would assume the aten convention.

> 2. should we let `wait()` appear in the IR? This might be related to how the cuda stream sync works in our current tracer.

Please see my other comments for the short term solution. Long term wise, yes we need a way to represent cuda streams in the graph. We don't know how yet.

wanchaol reviewed (approved):
Looks good to me, looks like the CI failure is real:
```
Broken ops: [
	c10d::broadcast(__torch__.torch.classes.c10d.ProcessGroup _0, Tensor[] _1, int _2, int _3, int _4) -> __torch__.torch.classes.c10d.Work _0
]
```
Could you fix the CI issue before landing? Thanks!

alanwaketan commented:
> Looks good to me, looks like the CI failure is real:
> 
> ```
> Broken ops: [
> 	c10d::broadcast(__torch__.torch.classes.c10d.ProcessGroup _0, Tensor[] _1, int _2, int _3, int _4) -> __torch__.torch.classes.c10d.Work _0
> ]
> ```
> 
> Could you fix the CI issue before landing? Thanks!

Thanks, Wanchao. I believe it's intended to break the schema. Do you know how to update the test expectation of the backward_compat test?

Commit:
Update on "[c10d] Make allreduce as a custom op"


Summary:
This patch makes allreduce as a custom op such that it's dispatcher
passable. It's one part of the effort to route comm ops to the dispatcher
such that tracing mechanisms that relies on the dispatcher can trace them,
e.g., LazyTensor and AOTAutograd.

Test Plan:
python test/distributed/test_c10d_nccl.py -k test_allreduce_ops
python test/distributed/test_c10d_gloo.py -k test_allreduce_basics
...and other existing distributed tests.

[ghstack-poisoned]

alanwaketan commented:
The XLA failure doesn't seem to be related.

alanwaketan commented:
@pytorchbot merge this

alanwaketan commented:
@pytorchbot merge -f

mrshenli commented:
> should we let wait() appear in the IR? This might be related to how the cuda stream sync works in our current tracer.

Can this be represented as edges in the graph?

alanwaketan commented:
> > should we let wait() appear in the IR? This might be related to how the cuda stream sync works in our current tracer.
> 
> Can this be represented as edges in the graph?

I think we need more discussions on this. Let me try to organize a follow up meeting.
